server/service/src/repack/generate.rs

```diff
@@ -0,0 +1,187 @@
+        item_name: item.name.clone(),
+        item_code: item.code.clone(),
+        stock_line_id: Some(new_stock_line.id.clone()),
+        location_id: new_stock_line.location_id.clone(),
+        batch: stock_line_to_update.batch.clone(),
+        expiry_date: stock_line_to_update.expiry_date,
+        pack_size: new_stock_line.pack_size,
+        r#type: InvoiceLineRowType::StockIn,
+        number_of_packs: new_stock_line.total_number_of_packs,
+        cost_price_per_pack: new_stock_line.cost_price_per_pack,
+        sell_price_per_pack: new_stock_line.sell_price_per_pack,
+        total_before_tax: new_stock_line.cost_price_per_pack
+            * new_stock_line.total_number_of_packs as f64,
+        total_after_tax: new_stock_line.cost_price_per_pack
+            * new_stock_line.total_number_of_packs as f64,
+        ..Default::default()
+    };
+
+    let stock_out = InvoiceLineRow {
+        id: uuid(),
+        stock_line_id: Some(stock_line_to_update.id.clone()),
+        location_id: stock_line_to_update.location_id.clone(),
+        pack_size: stock_line_to_update.pack_size,
+        r#type: InvoiceLineRowType::StockOut,
+        number_of_packs: number_of_packs_input,
+        cost_price_per_pack: stock_line_to_update.cost_price_per_pack,
+        sell_price_per_pack: stock_line_to_update.sell_price_per_pack,
+        total_before_tax: stock_line_to_update.cost_price_per_pack * number_of_packs_input as f64,
+        total_after_tax: stock_line_to_update.cost_price_per_pack * number_of_packs_input as f64,
+        ..stock_in.clone()
+    };
+
+    invoice_lines.push(stock_in);
+    invoice_lines.push(stock_out);
+
+    Ok((invoice, invoice_lines))
+}
+
+fn generate_new_stock_lines(stock_line: &StockLineRow, input: &InsertRepack) -> StockLineJob {
+    let stock_line_to_update = StockLineRow {
+        available_number_of_packs: stock_line.available_number_of_packs - input.number_of_packs,
+        total_number_of_packs: stock_line.total_number_of_packs - input.number_of_packs,
+        ..stock_line.clone()
+    };
+
+    let new_stock_line = {
+        let mut new_line = stock_line.clone();
+        let difference = input.new_pack_size as f64 / stock_line.pack_size as f64;
+
+        new_line.id = uuid();
+        new_line.pack_size = input.new_pack_size;
+        new_line.available_number_of_packs =
+            input.number_of_packs * stock_line.pack_size as f64 / input.new_pack_size as f64;
+        new_line.total_number_of_packs =
+            input.number_of_packs * stock_line.pack_size as f64 / input.new_pack_size as f64;
+        new_line.sell_price_per_pack = stock_line.sell_price_per_pack * difference;
+        new_line.cost_price_per_pack = stock_line.cost_price_per_pack * difference;
+        new_line.location_id = input.new_location_id.clone();
+
+        new_line
+    };
+
+    StockLineJob {
+        new_stock_line,
+        stock_line_to_update,
+    }
+}
+
+pub fn generate_location_movement(
```

Comment:
I still think we only need to generate one location movement (if location_id is provided), and it's for the new stock line, existing stock line is not being moved anywhere it's just being reduced.

Reply:
But some of the existing stock is being moved out, do we not need to record that it moved out from x location at x time?

Reply:
> do we not need to record that it moved out from x location at x time

I think it would be quite confusing for stock line to look like it's moved out, while still being at that location (just because part of it was split)
 
There is an issue to figure out how to keep track of historic movements of a stock line (before it was split), up here: https://github.com/openmsupply/open-msupply/issues/1432